src/ChannelServer/World/Inventory.cs

```diff
@@ -194,19 +204,71 @@ public Item GetItem(EquipSlot slot)
 		/// </summary>
 		/// <param name="item"></param>
 		/// <return>Index of the item.</return>
-		public int AddSilent(Item item)
+		public List<object> AddSilent(Item item)
 		{
 			var cat = item.Data.Category;
-
 			lock (_syncLock)
 			{
+				List<object> items_updated = new List<object>();
+
 				if (!_items.ContainsKey(cat))
 					throw new ArgumentException("Unknown item category.");
 
-				_items[cat].Add(item);
-				_itemsWorldIndex[item.WorldId] = item;
+				var subIndex = -1;
+				var expected_amount = item.Amount;
+
+				if (item.Data.MaxStack > 1 && this.HasItem(item.Id))
+				{
+					int item_index = subIndex;
+					while (expected_amount > 0)
+					{
+						long inventory_index = this.GetStackableItemIndex(cat, item.Id, item.Amount);
+						subIndex = (int)inventory_index;
+						if (subIndex != -1 && _items[cat].Count > subIndex)
+						{
+							var possible_add_to_current_stack = _items[cat][subIndex].Data.MaxStack - _items[cat][subIndex].Amount;
+
+							if(expected_amount > possible_add_to_current_stack)
+							{
+								_items[cat][subIndex].Amount += possible_add_to_current_stack;
+								_itemsWorldIndex[_items[cat][subIndex].WorldId] = _items[cat][subIndex];
+
+								expected_amount -= possible_add_to_current_stack;
+								item.Amount -= possible_add_to_current_stack;
+							}
+							else
+							{
+								possible_add_to_current_stack = expected_amount;
+								_items[cat][subIndex].Amount += possible_add_to_current_stack;
+								_itemsWorldIndex[_items[cat][subIndex].WorldId] = _items[cat][subIndex];
+
+								expected_amount -= possible_add_to_current_stack;
+								item.Amount -= possible_add_to_current_stack;
+							}
+
+							item_index = subIndex;
+							items_updated.Add(new { index = item_index, diff = possible_add_to_current_stack });
```

Comment:
Since you have two values, an index and a diff for the index, you basically have a key/value relation, which you could return as a Dictionary.